dandischema/datacite/__init__.py

```diff
@@ -65,17 +80,102 @@
 
 
+def construct_unvalidated_dandiset(meta_dict: dict) -> Dandiset:
+    """
+    Construct a Dandiset model from a dictionary without validation.
+
+    Process all nested data first, then construct the final object.
+    """
+    # Work on a copy to avoid mutating the input
+    processed_dict = deepcopy(meta_dict)
+
+    if "license" in processed_dict and processed_dict["license"]:
+        processed_dict["license"] = [
+            LicenseType(item) for item in processed_dict["license"]
+        ]
+
+    if "contributor" in processed_dict and processed_dict["contributor"]:
+        processed_contributors: List[Union[Person, Organization]] = []
+        for contributor_data in processed_dict["contributor"]:
+            if "roleName" in contributor_data and contributor_data["roleName"]:
+                contributor_data["roleName"] = [
+                    RoleType(role) for role in contributor_data["roleName"]
+                ]
+
+            schema_key = contributor_data.get("schemaKey")
+            if schema_key == "Person":
+                processed_contributors.append(
+                    Person.model_construct(**contributor_data)
+                )
+            elif schema_key == "Organization":
+                processed_contributors.append(
+                    Organization.model_construct(**contributor_data)
+                )
+
+        processed_dict["contributor"] = processed_contributors
+
+    if "relatedResource" in processed_dict and processed_dict["relatedResource"]:
+        processed_resources = []
+        for resource_data in processed_dict["relatedResource"]:
+            if "relation" in resource_data:
+                resource_data["relation"] = RelationType(resource_data["relation"])
+            processed_resources.append(Resource.model_construct(**resource_data))
+
+        processed_dict["relatedResource"] = processed_resources
+
+    return Dandiset.model_construct(**processed_dict)
+
+
 def to_datacite(
     meta: Union[dict, PublishedDandiset],
     validate: bool = False,
     publish: bool = False,
+    *,
+    event: Optional[str] = None,
 ) -> dict:
-    """Convert published Dandiset metadata to Datacite"""
-    if not isinstance(meta, PublishedDandiset):
-        meta = PublishedDandiset(**meta)
+    """
+    Convert Dandiset metadata to DataCite payload.
+
+    This function tries to validate the metadata against PublishedDandiset model.
+    If strict validation fails, it falls back to using construct_unvalidated_dandiset()
+    to build the model without validation but with properly handled nested types.
+    """
+    if isinstance(meta, dict):
+        meta = deepcopy(meta)
+        try:
+            meta = PublishedDandiset(**meta)
+        except ValidationError:
+            # mypy can't track that meta is still dict after failed PublishedDandiset(**meta)
+            assert isinstance(meta, dict)
+            if meta.get("version") == "draft":
```

Comment:
if the version is a draft, it wont have fields like datePublished. However this can also happen when we are creating a Dandiset DOI from a published version-- in this case, the metadata is the published version, but the doi and the url fields wont pass validation (they both won't include the version). 

Previously I modified the PublishedDandiset to accept either format for url and doi, but I dont think that really makes sense-- those aren't valid for a published dandiset, and we wouldnt want our output schema to reflect flexibility thats not really there.